exonum/src/api/mod.rs

```diff
@@ -106,37 +95,47 @@ impl From<io::Error> for ApiError {
     }
 }
 
-impl From<StorageError> for ApiError {
-    fn from(e: StorageError) -> ApiError {
+impl From<storage::Error> for ApiError {
+    fn from(e: storage::Error) -> ApiError {
         ApiError::Storage(e)
     }
 }
 
-impl From<FromHexError> for ApiError {
-    fn from(e: FromHexError) -> ApiError {
-        ApiError::FromHex(e)
-    }
-}
-
 impl From<ApiError> for IronError {
     fn from(e: ApiError) -> IronError {
-        use std::error::Error;
-
-        let mut body = BTreeMap::new();
-        body.insert("debug", format!("{:?}", e));
-        body.insert("description", e.description().to_string());
         let code = match e {
-            ApiError::FileExists(hash) |
-            ApiError::FileNotFound(hash) => {
-                body.insert("hash", encode_hex(&hash));
-                status::Conflict
+            ApiError::FileExists(..) => status::Conflict,
+            ApiError::FileNotFound(..) |
+            ApiError::NotFound => status::NotFound,
+            // Note that `status::Unauthorized` does not fit here, because
+            //
+            // > A server generating a 401 (Unauthorized) response MUST send a
+            // > WWW-Authenticate header field containing at least one challenge.
+            //
+            // https://tools.ietf.org/html/rfc7235#section-4.1
+            ApiError::Unauthorized => status::Forbidden,
+
+            ApiError::BadRequest(..) |
+            ApiError::FileTooBig => status::BadRequest,
+
+            ApiError::Storage(..) |
+            ApiError::Io(..) |
+            ApiError::InternalError(..) => status::InternalServerError,
+        };
+        let body = {
+            let mut map = BTreeMap::new();
+            map.insert("debug", format!("{:?}", e));
+            map.insert("description", e.to_string());
+            match e {
+                ApiError::FileExists(hash) |
+                ApiError::FileNotFound(hash) => {
+                    map.insert("hash", encode_hex(&hash));
+                }
+                _ => (),
             }
-            _ => status::Conflict,
+            ::serde_json::to_string_pretty(&map).unwrap()
```

Comment:
Is it only me, who doesn't like `::` prefixes? :smile:

Reply:
Yep, it definitely  is not required here! Although, I personally like absolute paths: you can understand them without context (that is, you don't need to scan use declarations)